Removal of Winston

briancullen opened the pull request:
Issue #, if available:
#10 #48 #103 

Description of changes:
Alternative approach to upgrading the logging to that proposed in #188 - both requests aren't needed, it's one or the other.

Instead of upgrading the Winston library this change creates a simple logger that can be used instead. This change also standardises the logging format for in all environments and ensures that the specified logging level is honoured in lambda environments.

As far as I can tell the meta (or second argument to the logging functions) is only used in two places, both of which log out err.stack. As previously implemented this was converted into a JSON string which made it hard to read. The logger was changed so that strings (such as the stack trace) are printed out without being stringified.

Additionally the ability to change the log level was added whereas previously this could only be done via environmental variables before the module is loaded.

Some additional error checking may be required before merging this request and I'm making this initial submission to show the alternative mentioned in #188 is feasible.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

willarmiros commented:
Hi @briancullen,
Thank you for yet another contribution. I would like to request that you make this PR against the `2.x` [branch](https://github.com/aws/aws-xray-sdk-node/tree/2.x) instead of `master`. See #192 for more details. We have been discussing removing winston for some time since it is overkill for our relatively simple use case, so I think we'd go with this PR instead of #188, so this action isn't necessary on your other PR.

Sorry for this inconvenience, we'll follow up with this PR as soon as this action is taken.

briancullen commented:
As requested by @willarmiros this pull request is being closed and replaced by #193 which is against the 2.x branch rather than master.

briancullen commented:
Hi @willarmiros 

Given the comments made about breaking changes on the associated commit in the 2.x branch I’m surprised this has been reopened. Do we no longer have to wait for the major release?

willarmiros commented:
Hi @briancullen,

We have come to a release plan for a 3.0.0-alpha version of the SDK that will be released from the master branch. Since many of our customers want this and other 3.x changes ASAP, we want to expedite the release process even though we have more breaking changes coming. I am currently reviewing this PR and after giving some feedback hope to have it merged very soon.

We can't merge #193 since we want to keep the 2.x branch completely clear of breaking changes, since it still represents major version 2.x of the SDK. Stay tuned for a post on #157 that will explain the whole 3.0.0 release process.

Very sorry for the confusion, but we thank you very much for your continued contributions and look forward to releasing this soon.

briancullen commented on packages/core/lib/logger.js:
```diff
@@ -1,96 +1,56 @@
+}
 
-      var result = [], prop, i;
+function formatLogMessage(level, message, meta) {
```
The fact that the time stamp is repeated for lambda did occur to me when I was writing this but won’t it be repeated for other things that write to cloudwatch as well? Say an ECS container for example?

At the time of writing I think I decided it was better to have a consistent log format that is produced regardless of the environment rather than trying to customise it for each different environment.

I’m happy to change it if that is the recommendation but, as I said, I don’t think this will avoid repeated timestamps in every situation.

An alternative could be to allow the user of the library to control the message format (either in total or just the inclusion of the time stamp) via another environment variable or api. I’m happy to prototype something along those lines if you want.

What do you think is the best approach?

willarmiros replied:
I think repeating the timestamp in CloudWatch is acceptable since CW treats their timestamp as a separate entity from the logging message, whereas Lambda actually interpolates the timestamp and log level into the logging message. I believe that is why the original author of the SDK special-cased logging in Lambda but not other environments.

If a user really cares about the logging format, they can always use the `setLogger()` API with their own logging object.

briancullen replied:
Changed so that if the AWS_LAMBDA_TASK environment variable is set then the timestamp and logging level is not included in the outputted log line.

briancullen commented on packages/core/lib/logger.js:
```diff
@@ -1,96 +1,56 @@
+  info: createLoggerForLevel('info'),
+  warn: createLoggerForLevel('warn'),
+  debug: createLoggerForLevel('debug'),
+  setLevel: (level) => logLevel = calculateLogLevel(level),
```
To take a step back, immediately after writing that set level method I kind of regretted it because I’m not entirely sure what the use case for it is. Particularly as the other methods are fairly comply implemented (take the winston logger for example) but setLevel isn’t.

So I suppose I’m asking do you think it is worth keeping?

willarmiros replied:
I agree that I can't really imagine a use case of needing to set the log level programmatically instead of with an EV. I have no issues with scrapping it.

briancullen replied:
Thanks for the feedback. Given how late it is here I’ll update the code as requested and push up a new version on or before Monday morning. Have a nice weekend.

briancullen replied:
Have removed the setLevel method as discussed so the environment variables are the only way to set the level of the logger.

briancullen commented on packages/core/lib/logger.js:
```diff
@@ -1,96 +1,56 @@
-      var result = [], prop, i;
+function formatLogMessage(level, message, meta) {
+  return createTimestamp() +' [' + level.toUpperCase() + '] ' +
+    (message !== undefined ? message : '') +
```
Changed as requested.

briancullen commented on packages/core/lib/logger.js:
```diff
@@ -1,96 +1,56 @@
-var logger;
-var xrayLogLevel = process.env.AWS_XRAY_LOG_LEVEL;
+var validLogLevels = [ 'trace', 'debug', 'info', 'warn', 'error'  ];
+var logLevel = calculateLogLevel(process.env.AWS_XRAY_DEBUG_MODE ? 'debug' : process.env.AWS_XRAY_LOG_LEVEL);
```
Have updated the readme to reflect the logging changes. As the getLevel and setLevel methods have been removed they are not mentioned.

briancullen commented on packages/core/lib/logger.js:
```diff
@@ -1,96 +1,56 @@
 
-var logger;
-var xrayLogLevel = process.env.AWS_XRAY_LOG_LEVEL;
+var validLogLevels = [ 'trace', 'debug', 'info', 'warn', 'error'  ];
```
Not sure why I put trace there either. It has been removed.

briancullen commented on packages/core/lib/logger.js:
```diff
@@ -1,96 +1,56 @@
+  getLevel: () => validLogLevels[logLevel]
+};
 
 /* eslint-disable no-console */
```
Changed as requested.

briancullen commented on packages/core/test/unit/logger.test.js:
```diff
@@ -9,35 +10,114 @@ describe('logger', function () {
+      assert.equal(logger.getLogger().getLevel(), 'error');
+    });
+
+    it('Should set logging level after initialisation', function() {
```
As setLevel, and it's associated tests, has been removed this change no longer applies.

briancullen commented:
Hello @willarmiros,

Think I addressed all the issues that you raised but let me know if I missed anything.

Apart from that if you do consider changing the log format to be a breaking change I suppose the last thin I'd like to check is that you are happy with the format of the logs produced particularly those that have metadata? Currently in the sdk code I think it is only stack traces you pass as metadata.
